components/brave_news/browser/feed_v2_builder.cc

```diff
@@ -485,6 +575,7 @@ void FeedV2Builder::BuildFeedFromArticles() {
   auto suggested_publisher_ids = suggested_publisher_ids_;
   auto articles = GetArticleInfos(raw_feed_items_, signals_);
   auto feed = mojom::FeedV2::New();
+  auto topics = topics_;
```

Comment:
This a full copy of the elements of `topics_` in to `topics`. Is that what we want?

Reply:
It is, yes - we copy the topics so we can mutate the copy to track which topics we've processed. At the moment, we just use them in order, so could probably keep track of which topic should be next with a `size_t`. However, in future this may change (maybe using Roulette wheel selection, or something similar).

Making the copy when building the feed means we're able to rebuild the feed without refetching topics from the server.

Reply:
But we actually don't change the elements themselves, do we? Unless I'm overlooking any other detail, we copy the topics to a local vector, then we gradually pop the first element of the vector, but we never actually mutate any topics. This seems to be a bit wasteful to me, because we are not only copying all these structs, we are also popping the first element every time we use one of the topics, which leads to all the remaining elements to be moved up inside the storage, and we also are doing an extra move, by moving the first element to a local variable.

I think we could have here something like:

```c++
auto topics = base::make_span(topics_);
```

This will take rather a reference to the topics in the vector, and then you could pass it around as `base::span<Topic>& topics`. 

In `GenerateTopicBlock`, rather than moving the first element out of the span, you could just take a reference to it like:

```c++
// this gets a reference to the first element but avoids copy.
auto& topic = topics.front();
```

And then once you are finished reading `topic`, by the end of `GenerateTopicBlock` you could shrink the span with:

```c++
topics = base::make_span(std::next(topics.begin()), topics.end());
```

This way we don't pay any copy or move ops both inside the vector, or when moving out the first element to the scope of `GenerateTopicBlock`.

Reply:
Thanks @cdesouza-chromium! That's much nicer!

Reply:
+1 here to use spans/views instead of moving/resizing vectors.

Such places quickly become bottlenecks when the amount of data suddenly changes.